flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/StreamArrowPythonGroupWindowAggregateFunctionOperator.java

```diff
@@ -0,0 +1,547 @@
+			if (triggerResult) {
+				emitWindowResult(window);
+			}
+			// register a clean up timer for the window
+			registerCleanupTimer(window);
+		}
+	}
+
+	@Override
+	@SuppressWarnings("ConstantConditions")
+	public void emitResult(Tuple2<byte[], Integer> resultTuple) throws Exception {
+		byte[] udafResult = resultTuple.f0;
+		int length = resultTuple.f1;
+		bais.setBuffer(udafResult, 0, length);
+		int rowCount = arrowSerializer.load();
+		for (int i = 0; i < rowCount; i++) {
+			Tuple2<RowData, W> input = inputKeyAndWindow.poll();
+			RowData key = input.f0;
+			W window = input.f1;
+			setWindowProperty(window);
+			windowAggResult.replace(key, arrowSerializer.read(i));
+			rowDataWrapper.collect(reuseJoinedRow.replace(windowAggResult, windowProperty));
+		}
+	}
+
+	@Override
+	public void onEventTime(InternalTimer<K, W> timer) throws Exception {
+		setCurrentKey(timer.getKey());
+
+		triggerContext.window = timer.getNamespace();
+		if (triggerContext.onEventTime(timer.getTimestamp())) {
+			// fire
+			emitWindowResult(triggerContext.window);
+		}
+
+		if (windowAssigner.isEventTime()) {
+			cleanWindowIfNeeded(triggerContext.window, timer.getTimestamp());
+		}
+	}
+
+	@Override
+	public void onProcessingTime(InternalTimer<K, W> timer) throws Exception {
+		setCurrentKey(timer.getKey());
+
+		triggerContext.window = timer.getNamespace();
+		if (triggerContext.onProcessingTime(timer.getTimestamp())) {
+			// fire
+			emitWindowResult(triggerContext.window);
+		}
+
+		if (!windowAssigner.isEventTime()) {
+			cleanWindowIfNeeded(triggerContext.window, timer.getTimestamp());
+		}
+	}
+
+	/**
+	 * Returns {@code true} if the watermark is after the end timestamp plus the allowed lateness
+	 * of the given window.
+	 */
+	private boolean isWindowLate(W window) {
+		return windowAssigner.isEventTime() &&
+			(cleanupTime(window) <= internalTimerService.currentWatermark());
+	}
+
+	/**
+	 * Returns the cleanup time for a window, which is
+	 * {@code window.maxTimestamp + allowedLateness}. In
+	 * case this leads to a value greated than {@link Long#MAX_VALUE}
+	 * then a cleanup time of {@link Long#MAX_VALUE} is
+	 * returned.
+	 *
+	 * @param window the window whose cleanup time we are computing.
+	 */
+	private long cleanupTime(W window) {
+		if (windowAssigner.isEventTime()) {
+			long cleanupTime = window.maxTimestamp() + allowedLateness;
+			return cleanupTime >= window.maxTimestamp() ? cleanupTime : Long.MAX_VALUE;
+		} else {
+			return window.maxTimestamp();
+		}
+	}
+
+	private void emitWindowResult(W window) throws Exception {
+		windowAccumulateData.setCurrentNamespace(window);
+		windowRetractData.setCurrentNamespace(window);
+		Iterable<RowData> currentWindowAccumulateDatas = windowAccumulateData.get();
+		if (currentWindowAccumulateDatas != null) {
+			currentBatchCount = 0;
+			for (RowData accumulateData : currentWindowAccumulateDatas) {
+				if (!hasRetractData(accumulateData)) {
```

Comment:
So for each input data, we have to iterate the retract state. I guess the performance should be very poor. Could we improve this?